Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts

```diff
@@ -0,0 +1,63 @@
+import { TaskParameters } from './taskparameters';
+import { KuduServiceUtility } from 'azurermdeploycommon/operations/KuduServiceUtility';
+import { AzureAppService } from 'azurermdeploycommon/azure-arm-rest/azure-arm-app-service';
+import { AzureRMEndpoint } from 'azurermdeploycommon/azure-arm-rest/azure-arm-endpoint';
+import { Kudu } from 'azurermdeploycommon/azure-arm-rest/azure-arm-app-service-kudu';
+import { AzureAppServiceUtility } from 'azurermdeploycommon/operations/AzureAppServiceUtility';
+import { AzureEndpoint } from 'azurermdeploycommon/azure-arm-rest/azureModels';
+import { AzureResourceFilterUtility } from 'azurermdeploycommon/operations/AzureResourceFilterUtility';
+import tl = require('vsts-task-lib/task');
+import { addReleaseAnnotation } from 'azurermdeploycommon/operations/ReleaseAnnotationUtility';
+import { ContainerBasedDeploymentUtility } from 'azurermdeploycommon/operations/ContainerBasedDeploymentUtility';
+
+export class AzureRmWebAppDeploymentProvider{
+    protected taskParams:TaskParameters;
+    protected appService: AzureAppService;
+    protected kuduService: Kudu;
+    protected appServiceUtility: AzureAppServiceUtility;
+    protected kuduServiceUtility: KuduServiceUtility;
+    protected azureEndpoint: AzureEndpoint;
+    protected activeDeploymentID;
+
+    constructor(taskParams: TaskParameters) {
+        this.taskParams = taskParams;
+    }
+
+    public async PreDeploymentStep() {
+        this.azureEndpoint = await new AzureRMEndpoint(this.taskParams.connectedServiceName).getEndpoint();
+        console.log(tl.loc('GotconnectiondetailsforazureRMWebApp0', this.taskParams.WebAppName));
+        
+        if(!this.taskParams.DeployToSlotOrASEFlag){
+            var appDetails = await AzureResourceFilterUtility.getAppDetails(this.azureEndpoint, this.taskParams.WebAppName);
+            this.taskParams.ResourceGroupName = appDetails["resourceGroupName"];
+        }
+
+        this.appService = new AzureAppService(this.azureEndpoint, this.taskParams.ResourceGroupName, this.taskParams.WebAppName, this.taskParams.SlotName);
+        this.appServiceUtility = new AzureAppServiceUtility(this.appService);
+
+        this.kuduService = await this.appServiceUtility.getKuduService();
+        this.kuduServiceUtility = new KuduServiceUtility(this.kuduService);
+        tl.debug(`Resource Group: ${this.taskParams.ResourceGroupName}`);
+    }
+
+    public async DeployWebAppStep() {
+        tl.debug("Performing container based deployment.");
+
+        let containerDeploymentUtility: ContainerBasedDeploymentUtility = new ContainerBasedDeploymentUtility(this.appService);
+        await containerDeploymentUtility.deployWebAppImage(this.taskParams);
```

Comment:
No much sure about the logic behind "We are waiting for each method to be executed before moving to next statement then what is logic of async method".  Why not sync ? We generally use Async method when we really don't main thread to block on it or context switch between thread to take care of it.

Reply:
You are correct we are doing asynchronous calls sequentially. But there is a difference with synchronous calls. With sync we block the thread until response is returned. With aync/await construct, the call returns a promise instantaneously but waits for it to settle down. Thus the thread is not blocked, and we still have a responsive UI. await ensures that the rest of the async function is executed once awaitable is completed.

Reply:
Here we are not using any UI and it is background task.